common/stats/stats.go

```diff
@@ -118,12 +107,89 @@ func (r *emaFixedWindowQPSTracker) Stop() {
 	r.wg.Wait()
 }
 
-func (r *emaFixedWindowQPSTracker) ReportCounter(delta int64) {
-	r.counter.Add(delta)
+func (r *emaFixedWindowQPSTracker) ReportCounter(amount int64) {
+	r.root.add(amount)
+}
+
+func (r *emaFixedWindowQPSTracker) ReportGroup(group string, amount int64) {
+	r.root.add(amount)
+	r.getOrCreate(group).add(amount)
+}
+
+func (r *emaFixedWindowQPSTracker) GroupQPS(group string) float64 {
+	state, ok := r.get(group)
+	if !ok {
+		return 0
+	}
+	return state.getQPS()
 }
 
 func (r *emaFixedWindowQPSTracker) QPS() float64 {
-	return r.qps.Load()
+	return r.root.getQPS()
+}
+
+func (r *emaFixedWindowQPSTracker) get(group string) (*emaFixedWindowState, bool) {
+	res, ok := r.groups.Load(group)
+	if !ok {
+		return nil, false
+	}
+	return res.(*emaFixedWindowState), true
+}
+
+func (r *emaFixedWindowQPSTracker) getOrCreate(group string) *emaFixedWindowState {
+	res, ok := r.groups.Load(group)
```

Comment:
nit: `groups` are never cleaned up. when an isolation group is gone (decommissioned zone) it will stay in the map. There is no unexpected behavior due to these leftover groups so it's a minor nit. We have other bigger gaps to make "isolation group auto discovery" a true restart-free feature.